### DISCUSSION TITLE
Support for ATtiny16x4 CPUs?

### DISCUSSION THREAD

**Discussion by l8gravely** [ukw100/IRMP#85] [General]
Hi all,
I'm looking to get some IR sending/receiving working with some AVR ATtiny16x4 CPUs and I was wondering if anyone had an hints on how to go about this?  What kind of linux based development setup would you suggest?   Especially just for testing the compilation and setup of the library for basic support. 

I'm currently using the Arduino 1.8.13 IDE( but hating it, want to move to simple emacs editing and makefiles) for my work.  

Right now I'm just hacking in the proper #ifdef ... #endif pairs to support __AVR_ATtiny1614__ and __AVR_ATtiny1604__.  

From the look of the code, I suspect I need to get the datasheet for the ATtiny1616 and compare it with the 16x4 datasheet to see how much different they really are.  Does this sound like a workable plan?  

Thanks,
John

**Comment by l8gravely** [comment]
So I've done the following attached patch as a first step to IRTimer.hpp, which fails with:

`wiring.c.o (symbol from plugin): In function `__vector_15':
(.text+0x0): multiple definition of `__vector_15'
/tmp/arduino_build_14346/sketch/ReceiveAndSend-attiny1614-test.ino.cpp.o (symbol from plugin):(.text+0x0): first defined here
/home/john/.arduino15/packages/DxCore/tools/avr-gcc/7.3.0-atmel3.6.1-azduino6/bin/../lib/gcc/avr/7.3.0/../../../../avr/bin/ld: Disabling relaxation: it will not work with multiple definitions
collect2: error: ld returned 1 exit status`


So I'm not sure what's the best way to get the pre-processor output dumped so I can try to trace through it.  I suspect there's a conflict somewhere, but it's not obvious since the ATtiny1616 and the ATtiny1614 (my target) are defined in the same datasheet, and are basically the same chip with differences in terms of pin count and such.   As you can see, I'm trying all kinds of hacks to see where I'm going wrong here.  

Any suggestions appreciated!  And yes, I've now also got some ATtiny1616 dev boards on order as well.  LOL!  But it would be nice to support more of the ATtiny 1/2 series MCUs if possible.   I find the way the IRTimer.hpp is setup is really ... opaque.  I almost think splitting it into seperate files would be the better method, since they you could just add in the right include and put in a new file.  Or maybe just breaking it down by family of processors.  Dunno... and I shouldn't complain, it looks like it's good software!

[diff.txt](https://github.com/IRMP-org/IRMP/files/12841786/diff.txt)

**Reply by l8gravely** [reply]
So... it came to me that I should test this code with an ATtiny1616 chip before I spent money on buying some test boards to complement the 1604 and 1614 MCUs I had at hand.  Silly me... I ordered some ATtiny1616 boards before I did my testing, and while the code and docs say they support ATtiny1616 chips, it fails with the same error as I pasted in above.  This is when just doing the "SimpleSender" example file, with out any changes.

**Reply by ArminJo** [reply]
Try [megaTinyCore](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/ATtiny_x14.md)
